File: Leetcode/1857 Largest Color Value in a Directed Graph.py
UN_VISITED, VISITING, VISITED = 0, 1, 2
N_COLORS = 26
CONTIAINS_CYCLE = -1

class Solution:
    def largestPathValue(self, colors, edges):
        self.colors = colors
        n = len(colors)
        self.adj_list = [[] for _ in range(n)]
        for src, dst in edges:
            self.adj_list[src].append(dst)

        self.visit_status = [UN_VISITED] * n
        self.color_frequencies = [[0] * N_COLORS for _ in range(n)]
        
        max_freq = 0
        for node in range(n):
            freq = self.dfs(node)
            if freq == CONTIAINS_CYCLE:
                return CONTIAINS_CYCLE
            max_freq = max(max_freq, freq)
        
        return max_freq

    def dfs(self, start):
        if self.visit_status[start] == VISITING:
            return CONTIAINS_CYCLE

        if self.visit_status[start] == VISITED:
            return max(self.color_frequencies[start])
        
        self.visit_status[start] = VISITING
        node_color = ord(self.colors[start]) - ord('a') 

        for neighbour in self.adj_list[start]:
            if self.dfs(neighbour) == CONTIAINS_CYCLE:
                return CONTIAINS_CYCLE
            
            for color in range(N_COLORS):
                self.color_frequencies[start][color] = max(
                    self.color_frequencies[start][color],
                    self.color_frequencies[neighbour][color]
                )

        self.color_frequencies[start][node_color] += 1
        self.visit_status[start] = VISITED
        return max(self.color_frequencies[start])
      

File: Leetcode/1857 Largest Color Value in a Directed Graph.ts
const UN_VISITED = 0, VISITING = 1, VISITED = 2;
type VisitStatus = 0 | 1 | 2;
const CONTAINS_CYCLE = -1;
const N_COLORS = 26;

function largestPathValue(colors: string, edges: number[][]): number {
    const n = colors.length;
    const adjList: number[][] = new Array();
    const frequentColorFreq: number[][] = new Array(n);
    for(let i = 0;i < n; i++) {
        adjList[i] = [];
        frequentColorFreq[i] = new Array(N_COLORS).fill(0);
    }
    for(const [src, dst] of edges) {
        adjList[src].push(dst);
    }
    const visitStatus: VisitStatus[] = new Array(n).fill(UN_VISITED);
    let mostFrequentFreq = -1;
    for(let node = 0; node < n; node++) {
        if(visitStatus[node] !== UN_VISITED) {
            continue;
        }
        const freqFromNode = dfs(node, adjList, colors, visitStatus, frequentColorFreq);
        if(freqFromNode === CONTAINS_CYCLE) {
            return -1;
        }
        mostFrequentFreq = Math.max(mostFrequentFreq, freqFromNode);
    }
    return mostFrequentFreq;
}

function dfs(
    start: number, 
    adjList: number[][],
    colors: string, 
    visitStatus: VisitStatus[], 
    frequentColorFreq: number[][]
): number {
    if (visitStatus[start] === VISITING) {
        return CONTAINS_CYCLE;
    }
    if (visitStatus[start] === VISITED) {
        return Math.max(...frequentColorFreq[start]);
    }
    visitStatus[start] = VISITING;
    const startColor = colors.charCodeAt(start) - 'a'.charCodeAt(0);
    for(const neighbour of adjList[start]) {
        const neighbourFreq = dfs(neighbour, adjList, colors, visitStatus, frequentColorFreq);
        if(neighbourFreq === CONTAINS_CYCLE) {
            return CONTAINS_CYCLE;
        }
        for(let color = 0; color < N_COLORS; color++) {
            frequentColorFreq[start][color] = Math.max(
                frequentColorFreq[start][color],
                frequentColorFreq[neighbour][color]
            );
        }
    }
    frequentColorFreq[start][startColor]++;
    visitStatus[start] = VISITED;
    return Math.max(...frequentColorFreq[start]);
}
